### DISCUSSION TITLE
Compiling vcpkg librairies with different versions of C++ (CMake)

### DISCUSSION THREAD

**Discussion by generaltoto** [microsoft/vcpkg#40309] [Q&A] [answered]
I have an application that uses multiple libraries imported/compiled via vcpkg. The project is then built using cmake in C++14 using `set(CMAKE_CXX_STANDARD 14)` since Activemq-cpp cannot be built with higher cpp versions.

My question is, is it possible to tell vcpkg/cmake to build (or link ?) activemq-cpp using C++14 and then set the C++ version of the final project to C++20?

_This might be more of a CMake question but I wondered if the problem was solvable using the vcpkg config files_ 

Here are the files:

### vcpkg.json
```json
{
  "dependencies": [
    {
      "name": "opencascade",
      "version>=": "7.8.1"
    },
    {
      "name": "protobuf",
      "version>=": "3.21.8"
    },
    {
      "name": "activemq-cpp",
      "version>=": "3.9.5#15"
    }
  ],
  "version-string": "0.13.1",
  "name": "opencascade-rendering-viewer"
}
```

### CMakeList.txt
```cmake
# --- General CMake settings -------------------------------------------
cmake_minimum_required(VERSION 3.21 FATAL_ERROR)


set(PROJECT_NAME "OpenCASCADE_Rendering_Server")
set(PROJECT_VERSION "0.13.1")
message(STATUS "Project Name: ${PROJECT_NAME} | Version: ${PROJECT_VERSION}")

project(${PROJECT_NAME} LANGUAGES CXX VERSION ${PROJECT_VERSION})

# Note from 26/07/2024
# This project uses Activemq-CPP, which cannot be build with c++ versions above 14 currently.
set(CMAKE_CXX_STANDARD 14)


# --- C++ Preprocessor Definitions --------------------------------------
add_compile_definitions($<$<CONFIG:DEBUG>:_DEBUG> $<$<NOT:$<CONFIG:DEBUG>>:NDEBUG>)
if (CMAKE_HOST_WIN32)
    add_compile_definitions(WIN32)
elseif (CMAKE_HOST_UNIX)
    add_compile_definitions(UNIX)
endif (CMAKE_HOST_WIN32)


# --- Packages ----------------------------------------------------------
find_package(OpenCASCADE CONFIG REQUIRED)

find_package(Protobuf CONFIG REQUIRED)

find_package(activemq-cpp CONFIG REQUIRED)


# --- Add custom CMake modules ------------------------------------------
include(cmake/protobufcompile.cmake)


# --- Add source files & Executable -------------------------------------
list(APPEND SRC /* Application .cpp files */)

list(APPEND HDR /* Application .h files */)

configure_file(config.h.in ${CMAKE_CURRENT_SOURCE_DIR}/src/config.h @ONLY)
add_executable(${PROJECT_NAME} ${SRC} ${HDR} ${PROTOBUF_GENERATED_FILES})

# --- Add include directories -------------------------------------------
target_include_directories(${PROJECT_NAME} PRIVATE src)


# --- Add external libraries main executable ----------------------------
target_link_libraries(${PROJECT_NAME} PRIVATE ${OpenCASCADE_LIBRARIES})
target_include_directories(${PROJECT_NAME} SYSTEM PRIVATE ${OpenCASCADE_INCLUDE_DIR})

target_link_libraries(${PROJECT_NAME} PRIVATE protobuf::libprotoc protobuf::libprotobuf protobuf::libprotobuf-lite)
target_include_directories(${PROJECT_NAME} SYSTEM PRIVATE ${PROTOBUF_COMPILER_CPP_OUT})

target_link_libraries(${PROJECT_NAME} PRIVATE unofficial::activemq-cpp::apr unofficial::activemq-cpp::activemq-cpp)
if (CMAKE_HOST_WIN32)
    target_link_libraries(${PROJECT_NAME} PRIVATE unofficial::activemq-cpp::ws2 unofficial::activemq-cpp::rpcrt4 unofficial::activemq-cpp::mswsock)
endif (CMAKE_HOST_WIN32)
```

Thanks for your time and let me know if you have more questions.

**Comment by Osyotr** [comment]
At the point where you `set(CMAKE_CXX_STANDARD 14)` `activemq-cpp` is already built. Vcpkg does not usually enforce specific C++ version while building ports and does not use your project's C++ version.
There might be issues when you *consume* a library if it has incompatible stuff like `auto_ptr` or `registry` in its public headers.

**Reply by generaltoto** [reply]
Yeah this is my problem, when I `set(CMAKE_CXX_STANDARD 20)`, I get a lot of errors about missing identifiers like `auto_ptr`.

So that would mean no possible use of `activemq-cpp` with C++ versions above 14 ?

**Reply by Osyotr** [reply]
I would just patch it. ¯⁠\⁠_⁠(⁠ツ⁠)⁠_⁠/⁠¯

**Reply by generaltoto** [reply]
Wdym ? Like removing all deprecated elements from the library ?

**Reply by Osyotr** [reply]
At least from its public interface, yeah.

**Reply by dg0yt** [reply]
Maybe the isolation is easier to maintain with separate libs within your project?

**Reply by generaltoto** [reply]
> Maybe the isolation is easier to maintain with separate libs within your project?

Not sure to understand, what do you mean by "_separate libs within your project_" ?

**Reply by dg0yt** [reply, marked as answer]
Instead of non-trivial patching the external activemq-cpp source code, separate your project code into one C++14 (`CXX_STANDARD` target property) lib which accesses pristine activemq-cpp, but doesn't expose the interfaces removed in modern C++ versions, and other libs which access *your* C++14 lib, but are not restricted themselves (`CMAKE_CXX_STANDARD`).

(I would assume you will eventually need to port away from activemq-cpp when they are *not*  making new releases. The separation of code could be aligned with that perspective.)

**Reply by generaltoto** [reply]
> Instead of non-trivial patching the external activemq-cpp source code, separate your project code into one C++14 (`CXX_STANDARD` target property) lib which accesses pristine activemq-cpp, but doesn't expose the interfaces removed in modern C++ versions, and other libs which access _your_ C++14 lib, but are not restricted themselves (`CMAKE_CXX_STANDARD`).

Ok I see now, I don't have the time to work on that for now but might do it later on if I really need some new C++ features. 
**Thank you guys for the tips.** I'm marking this conversation resolved for now. Will come back to it if I need help implementing what we talked about.
